### ISSUE TITLE
NUMERIC filter appears to stop working

### ISSUE THREAD

**Issue by timmixell** [RediSearch/RediSearch#1085]
Greetings,

I posted details and database on the Google Group: https://groups.google.com/forum/#!topic/redisearch/7Ev4rlDwYHU, but as I've continued to poke this particular behavior, I believe it to be either a glaring oversight on my part, or an issue with numeric indexing in Redisearch.

Against a populated index, if I perform a `REPLACE PARTIAL`:

`127.0.0.1:6379> ft.add activeteams:20190730 testing_teamID_5278 1 REPLACE PARTIAL FIELDS eventID 705`

And then:

```
127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5278 5278]'
1) (integer) 1
2) "testing_teamID_5278"
```

I can run that command successfully for a little bit, and then the search just stop returning results until I "nudge" the document again. What's bizarre about this, is that I can perform a ranged search and find the document after the "equals" search stops returning results:

```
127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[1 5278]' sortby teamID desc limit 0 1
1) (integer) 263
2) "testing_teamID_5278"
```

I've watched the output from `MONITOR` in another console window, and nothing is catching my eye, nor am I seeing any relevant information in the logs. What am I seeing here? Please let me know how I may be of assistance.

Thank you!

**Comment by MeirShpilraien** [comment]
Can you specify redisearch version you are running with? Can you test with 1.6.8?

Also can you give us an rdb that reproduce the issue?

**Comment by MeirShpilraien** [comment]
Never mind did not see you already gave an rdb here: https://groups.google.com/forum/#!topic/redisearch/7Ev4rlDwYHU

**Comment by timmixell** [comment]
Greetings @MeirShpilraien , 

Running:
Redis 5.0.7 (00000000/0) 64 bit
RediSearch version 1.6.8 (Git=v1.6.8)

I can still reproduce the behavior outlined above against my "teams" index:

```
127.0.0.1:6379> ft.add activeteams:20190730 testing_teamID_5083 1 REPLACE PARTIAL FIELDS eventID 582
OK

127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5083 5083]'
1) (integer) 1

[ repeat a handful of times ]

127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5083 5083]'
1) (integer) 0
```

I originally spotted this behavior on a Sentinel cluster. I've since pared back to a vanilla Redis server, with the same result. I can reproduce searching against either `teamID` (5083 in this case), and `eventID` (582).

Let me know if I can offer anything else. Thank you!

**Comment by MeirShpilraien** [comment]
@timmixell I just tried it on you rdb and it works:
```
127.0.0.1:6379> ft.add activeteams:20190730 testing_teamID_5083 1 REPLACE PARTIAL FIELDS eventID 582 teamID 5083
OK
127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5083 5083]'
1) (integer) 1
2) "testing_teamID_5083"
3) 1) "eventID"
   2) "582"
   3) "teamID"
   4) "5083"
127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5083 5083]'
1) (integer) 1
2) "testing_teamID_5083"
3) 1) "eventID"
   2) "582"
   3) "teamID"
   4) "5083"
127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5083 5083]'
1) (integer) 1
2) "testing_teamID_5083"
3) 1) "eventID"
   2) "582"
   3) "teamID"
   4) "5083"
```

Are you sure no one is deleting this document? Also get you check if the document exists using `hgetall testing_teamID_5083`

**Comment by timmixell** [comment]
Interesting. I could reproduce on two separate servers. The second being a one that only I was working with. I should have been more clear in my steps to repro: I noticed it would occasionally happen after 3-4 search attempts, but always within 10-15 seconds of the ft.add.

As I mentioned above, range searches continue to work, and the document is definitely still there, its just the "equals" search that stops returning results.

If you still can't reproduce, do you have any suggestions on how I might isolate this issue?

**Comment by timmixell** [comment]
I need to apologize: when I spun up the separate server to test, I seeded directly from my source of truth, which had a smaller result set than what I shared in the dump.rdb. In my local instance, `teamID` 5083 is the biggest ID in that range, while in the file I shared, ID 5278 is (5083 wasn't in the original dump.rdb I shared, so it was upserted).

It appears that I can only consistently reproduce this when working with numeric fields on the higher range of values expected in that field.

Confirmed against the dump.rdb shared in Google Group, I can do the following:

```
127.0.0.1:6379> ft.add activeteams:20190730 testing_teamID_5278 1 REPLACE PARTIAL FIELDS eventID 705
OK
127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5278 5278]'
1) (integer) 1
2) "testing_teamID_5278"
...

[ 3 searches later ]

127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5278 5278]'
1) (integer) 0
127.0.0.1:6379> ft.get activeteams:20190730 testing_teamID_5278
 1) "eventID"
 2) "705"
...
 ```

If I modify a record on the low-end of that numeric range (ID 5000), I cannot reproduce this issue:
```
127.0.0.1:6379> ft.add activeteams:20190730 testing_teamID_5000 1 REPLACE PARTIAL FIELDS eventID 705
OK
127.0.0.1:6379> ft.search activeteams:20190730 '@teamID:[5000 5000]'
1) (integer) 1
2) "testing_teamID_5000"
...
```
In the rdb I shared, the behavior starts to manifest reliably somewhere after teamID 5025.

I apologize again for the mix up.

**Comment by MeirShpilraien** [comment]
Do you know if from the time you get results to the time you don't any other documents was added to the database?

**Comment by timmixell** [comment]
The only document modification/access occurring are the ft.add/ft.search commands I've referenced above.

**Comment by timmixell** [comment]
I've continued to dig, and I believe there to be an issue with `GC_POLICY FORK`. I cannot reproduce the issue with `NOGC`, nor was I able to with `LEGACY` (thought I don't know how to force a garbage collection here). I had previously been running without any GC configuration (shame on me), but the concern would be the impact any time garbage collection occurs. 

Running:
Redis 5.0.7 (00000000/0) 64 bit
RediSearch version 1.6.8 (Git=v1.6.8)
`loadmodule /opt/redisearch/src/redisearch.so GC_POLICY FORK FORK_GC_RUN_INTERVAL 10 FORK_GC_CLEAN_THRESHOLD 10`

The problem manifests with the following:

```
FLUSHDB

FT.CREATE issue1085 SCHEMA foo TEXT SORTABLE bar NUMERIC SORTABLE

FT.ADD issue1085 document_0 1 REPLACE FIELDS foo 'foo 0' bar 0
FT.ADD issue1085 document_1 1 REPLACE FIELDS foo 'foo 1' bar 1
FT.ADD issue1085 document_2 1 REPLACE FIELDS foo 'foo 2' bar 2
FT.ADD issue1085 document_3 1 REPLACE FIELDS foo 'foo 3' bar 3
FT.ADD issue1085 document_4 1 REPLACE FIELDS foo 'foo 4' bar 4
FT.ADD issue1085 document_5 1 REPLACE FIELDS foo 'foo 5' bar 5
FT.ADD issue1085 document_6 1 REPLACE FIELDS foo 'foo 6' bar 6
FT.ADD issue1085 document_7 1 REPLACE FIELDS foo 'foo 7' bar 7
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_9 1 REPLACE FIELDS foo 'foo 9' bar 9

FT.SEARCH issue1085 '@bar:[8 8]'

FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8
FT.ADD issue1085 document_8 1 REPLACE FIELDS foo 'foo 8' bar 8

FT.SEARCH issue1085 '@bar:[8 8]'
```

**Comment by MeirShpilraien** [comment]
Thanks @timmixell its really help, I will check if I can reproduce locally.

**Comment by MeirShpilraien** [comment]
@timmixell thanks for reporting, I could reproduce and found the underline problem. It was really a bug in the forkgc that update the numeric tree incorrectly.
Fix is on this PR: https://github.com/RediSearch/RediSearch/pull/1089

**Comment by timmixell** [comment]
Thank you! Sorry for the initial confusion on how to reproduce.

**Comment by MeirShpilraien** [comment]
@timmixell Can we close this issue? it's been fixed on 1.6.10.

**Comment by timmixell** [comment]
👍 sorry for delay!
